### PR TITLE
Adding Drivenets vendor

### PR DESCRIPTION
**Pull request by davnerson-dn**

### CODE REVIEW

File: deploy/deploy.go
Comment (alexmasi) [inline deploy/deploy.go:1144]: please remove this deprecated field

File: deploy/deploy.go
Comment (alexmasi) [inline deploy/deploy.go:1154]: please add unit tests in deploy_test.go

File: deploy/kne/external-multinode.yaml
Comment (alexmasi) [inline deploy/kne/external-multinode.yaml:31]: please remove from these examples, instead feel free to create a new example (kind-bridge-cdnos.yaml or similar)

File: docs/create_topology.md
Comment (alexmasi) [inline docs/create_topology.md:128]: this can be removed after the field is removed

File: examples/drivenets/srlinux.cfg
Comment (alexmasi) [inline examples/drivenets/srlinux.cfg:1]: prefer just copying the config here in case paths change in future

File: proto/controller.proto
Comment (alexmasi) [inline proto/controller.proto:136]: remove

File: proto/topo.proto
Comment (alexmasi) [inline proto/topo.proto:67]: remove, this field is deprecated and no longer used, actually a good reminder for me to just delete this entirely

File: topo/node/drivenets/drivenets.go
Comment (alexmasi) [inline topo/node/drivenets/drivenets.go:214]: this should probably point to a moving image tag, is this open source btw?

File: topo/node/drivenets/drivenets.go
Comment (alexmasi) [inline topo/node/drivenets/drivenets.go:255]: does this support p4rt as well?

File: deploy/deploy.go
Comment (davnerson-dn) [reply deploy/deploy.go:1154]: unit test added

File: deploy/kne/external-multinode.yaml
Comment (davnerson-dn) [reply deploy/kne/external-multinode.yaml:31]: removed from examples and created new examples:
`kind-bridge-cdnos.yaml`
`external-multinode-cdnos.yaml`

File: docs/create_topology.md
Comment (davnerson-dn) [reply docs/create_topology.md:128]: removed

File: proto/controller.proto
Comment (davnerson-dn) [reply proto/controller.proto:136]: removed

File: proto/topo.proto
Comment (davnerson-dn) [reply proto/topo.proto:67]: removed

File: topo/node/drivenets/drivenets.go
Comment (davnerson-dn) [reply topo/node/drivenets/drivenets.go:214]: what do you mean by "a moving immage tag" ?
no the image is not open source, need to contact Drivenets

Comment (chrisy) [reply topo/node/drivenets/drivenets.go:214]: The default image tag should look like it points to a production release, I imagine something like `whatever.com/cdnos:latest`, even if that is not publicly accessible -- the one shown looks like it's an internal development build. For that dev work, you can always override the image in the config section of the node in the topology with the `image` statement.

IMO, the reason for wanting a more "final" looking image tag is that a user is more likely to have such an image tag loaded into the container runtime than the dev build one, and `:latest` is the convention for finding the most recent one.

I, as the eventual customer of this image, could similarly use that `image` statement to point KNE at a copy of a specific version of the cDNOS image in our internal registry.

File: examples/drivenets/srlinux.cfg
Comment (davnerson-dn) [reply examples/drivenets/srlinux.cfg]: ok, i copied the config

File: topo/node/drivenets/drivenets.go
Comment (davnerson-dn) [reply topo/node/drivenets/drivenets.go]: ok, understood, i setup a more official path

File: topo/node/drivenets/drivenets.go
Comment (davnerson-dn) [reply topo/node/drivenets/drivenets.go:255]: no, we don't support p4rt

**Comment by davnerson-dn** [comment]
i applied `go mod tidy` and `go fmt` ,
how to retrigger the presubmit ?

**Comment by alexmasi** [comment]
The presubmits should trigger automatically after your first PR, sorry

File: topo/node/drivenets/drivenets_test.go
Comment (alexmasi) [inline topo/node/drivenets/drivenets_test.go:44]: this could be streamlined quite a bit using the errdiff package, see other examples in this repo: https://github.com/openconfig/kne/blob/f197e16f84e47b8eadc4263d71fafe92b28dfbd9/exec/run/run_test.go#L108

File: topo/node/drivenets/drivenets.go
Comment (alexmasi) [inline topo/node/drivenets/drivenets.go:291]: can flip this conditional to reduce indentation:

```
if data == nil {
    return nil, nil
}
...
```

File: proto/controller.proto
Comment (alexmasi) [inline proto/controller.proto:136]: should be `= 1`

File: manifests/controllers/cdnos/manifest.yaml
Comment (alexmasi) [inline manifests/controllers/cdnos/manifest.yaml:579]: is this open source?

File: examples/drivenets/cdnos.pb.txt
Comment (alexmasi) [inline examples/drivenets/cdnos.pb.txt:6]: config_file and config_path should be defaulted to simplify topo (if these are standard values), see other vendors node impls

**Review by alexmasi — CHANGES_REQUESTED** [review]

**Comment by alexmasi** [comment]
Okay looks like the presubmits were addressed, I've left my last wave of review comments. After those are addressed, we should be good to merge.

File: manifests/controllers/cdnos/manifest.yaml
Comment (davnerson-dn) [reply manifests/controllers/cdnos/manifest.yaml:579]: yes, see  here :
https://github.com/drivenets/cdnos-controller

File: topo/node/drivenets/drivenets.go
Comment (davnerson-dn) [reply topo/node/drivenets/drivenets.go]: ok, i modifed the code accordingly

File: proto/controller.proto
Comment (davnerson-dn) [reply proto/controller.proto:136]: done

File: topo/node/drivenets/drivenets_test.go
Comment (davnerson-dn) [reply topo/node/drivenets/drivenets_test.go]: i modifed to use `errdif`

File: examples/drivenets/cdnos.pb.txt
Comment (davnerson-dn) [reply examples/drivenets/cdnos.pb.txt]: added default config file and path for node

**Review by alexmasi — APPROVED** [review]

**Comment by alexmasi** [comment]
LGTM, will merge assuming the kne-presubmit passes. It appears the linter is currently broken so will not block on that failure.

**Comment by alexmasi** [comment]
Thanks David for the contribution!

**Pull request merged**
